cpp/src/randomforest/randomforest_impl.cuh

```diff
@@ -169,48 +178,93 @@ const DecisionTree::DecisionTreeClassifier<T>* rfClassifier<T>::get_trees_ptr()
  * @param[in] forest: CPU point to RandomForestMetaData struct.
  */
 template <typename T>
-void rfClassifier<T>::fit(const cumlHandle& user_handle, T* input, int n_rows,
-                          int n_cols, int* labels, int n_unique_labels,
+void rfClassifier<T>::fit(const cumlHandle& user_handle, const T* input,
+                          int n_rows, int n_cols, int* labels,
+                          int n_unique_labels,
                           RandomForestMetaData<T, int>*& forest) {
   this->error_checking(input, labels, n_rows, n_cols, false);
 
   int n_sampled_rows = this->rf_params.rows_sample * n_rows;
+  int n_streams = this->rf_params.n_streams;
 
   const cumlHandle_impl& handle = user_handle.getImpl();
   cudaStream_t stream = user_handle.getStream();
-
+  cumlHandle local_handle[n_streams];
```

Comment:
We would ideally not want to create/destroy any cuda-related data structures as well as cumlHandle inside libcuml++. We should instead be using the `cumlHandle_impl::getInternalStream` to obtain "internal" streams for scheduling work.

Currently, the number of such internal streams is fixed. That can be a restriction for you since `n_streams` is a runtime option here. So, we could try to extend `cumlHandle/cumlHandle_impl` APIs to support runtime configuration of this.

Reply:
Tagging @jirikraus and @cjnolet for further discussions.

Reply:
One thing that can help you is having `TemporaryMemory` accept an explicit stream parameter, alongside with `cumlHandle_impl`. This decoupling of params will then help you use the same handle object across multiple TemporaryMemory's and still work on different streams.

Reply:
yes I think in that case, we should extend the option on how many internal streams. yes we can explicitly just create cudaStream and pass it to temporary memory of RF. This was we can use same handle with different streams.

Given that we need new functionality in cuml_handle. What would you recommend in mean time ?

Reply:
Let me check and get back to you on this one...

Reply:
I agree these streams, like the other resources, should be managed by the `cumlHandle`. Perhaps we can move the `n_streams` parameter to the `cumlHandle` constructor (maybe rename it to `n_internal_streams` and make it default to `1`). What do you guys think?

Also, since the release is right around the corner, what do you say we make a ticket for this and do it at the very beginning of 0.10?

Reply:
Makes sense. For now, I have filed issue #931 to track this one. Let's leave this as-it-is in this PR, atleast.